dask/order.py

```diff
@@ -381,6 +381,75 @@ def process_runnables() -> None:
 
+    # Only build this mapping if it's needed. This can be expensive for
+    # all-to-all communications, but we normally don't need it in this
+    # case
+    occurences_grouped_sorted: dict[int, list[Key]] = {}
+
+    def _built_occurrences_group_sorted() -> None:
+        if occurences_grouped_sorted:
+            return
+        occurrences: defaultdict[Key, int] = defaultdict(int)
+        for t in leaf_nodes:
+            for r in roots_connected[t]:
+                occurrences[r] += 1
+        occurences_grouped = defaultdict(set)
+        for root, occ in occurrences.items():
+            occurences_grouped[occ].add(root)
+        for k, v in occurences_grouped.items():
+            occurences_grouped_sorted[k] = sorted(v, key=sort_key, reverse=True)
+
+    def pick_seed() -> Key | None:
+        if not occurences_grouped_sorted:
+            _built_occurrences_group_sorted()
+
+        while occurences_grouped_sorted:
+            key = max(occurences_grouped_sorted)
+            picked_root = occurences_grouped_sorted[key][-1]
+            if picked_root in result:
+                occurences_grouped_sorted[key].pop()
+                if not occurences_grouped_sorted[key]:
+                    del occurences_grouped_sorted[key]
+                continue
+            return picked_root
+        return None
+
+    def get_target(longest_path: bool = False) -> Key:
+        # Some topologies benefit if the node with the most dependencies
+        # is used as first choice, others benefit from the opposite.
+        candidates = leaf_nodes
+        skey: Callable = sort_key
+
+        preferred_candidates = set()
+        if runnable_hull:
+            skey = lambda k: (num_needed[k], sort_key(k))
+            preferred_candidates = runnable_hull & candidates
+        if not preferred_candidates and reachable_hull:
+            skey = lambda k: (num_needed[k], sort_key(k))
+            preferred_candidates = reachable_hull & candidates
+
+        if reachable_hull or runnable_hull:
+            # We can't reach a leaf node directly, but we still have nodes
+            # with results in memory, these notes can inform our path towards
+            # a new preferred leaf node.
+            if not preferred_candidates:
+                hull = runnable_hull if runnable_hull else reachable_hull
+                for c in hull:
+                    preferred_candidates.update(leafs_connected[c])
+            candidates = preferred_candidates
+
+        if not candidates:
+            if seed := pick_seed():
+                candidates = leafs_connected[seed]
+            else:
+                candidates = runnable_hull or reachable_hull
+
+        # FIXME: This can be very expensive
+        if longest_path and not runnable_hull and not reachable_hull:
+            return leaf_nodes_sorted.pop()
+        else:
+            return min(candidates, key=skey)
+
     def use_longest_path() -> bool:
```

Comment:
I had to add this back in. The non-longest path approach is really expensive for embarrassingly parallel things